lightningd/onchain_control.c

```diff
@@ -692,33 +715,63 @@ static struct bitcoin_tx *onchaind_tx(const tal_t *ctx,
 	/* Worst-case sig is 73 bytes */
 	weight = bitcoin_tx_weight(tx) + 1 + 3 + 73 + 0 + tal_count(info->wscript);
 	weight += elements_tx_overhead(chainparams, 1, 1);
-	fee = amount_tx_fee(feerate, weight);
-
-	/* Result is trivial?  Spend with small feerate, but don't wait
-	 * around for it as it might not confirm. */
-	if (!amount_sat_add(&min_out, channel->our_config.dust_limit, fee))
-		fatal("Cannot add dust_limit %s and fee %s",
-		      type_to_string(tmpctx, struct amount_sat, &channel->our_config.dust_limit),
-		      type_to_string(tmpctx, struct amount_sat, &fee));
-
-	if (amount_sat_less(out_sats, min_out)) {
-		/* FIXME: We should use SIGHASH_NONE so others can take it? */
-		/* Use lowest possible theoretical fee: who cares if it doesn't propagate */
-		fee = amount_tx_fee(FEERATE_FLOOR, weight);
-		*worthwhile = false;
-	} else
-		*worthwhile = true;
-
-	/* This can only happen if FEERATE_FLOOR is still too high; shouldn't
-	 * happen! */
-	if (!amount_sat_sub(&amt, out_sats, fee)) {
-		amt = channel->our_config.dust_limit;
-		log_broken(channel->log, "TX can't afford minimal feerate"
-			   "; setting output to %s",
-			   type_to_string(tmpctx, struct amount_sat,
-					  &amt));
-		*worthwhile = false;
+
+	block_target = info->deadline_block;
+	for (;;) {
+		struct amount_sat fee;
+		u32 feerate;
+
+		feerate = feerate_for_target(ld->topology, block_target);
+		fee = amount_tx_fee(feerate, weight);
+
+		log_debug(channel->log,
+			  "Feerate for target %"PRIu64" (%+"PRId64" blocks) is %u, fee %s of %s",
+			  block_target,
+			  block_target - get_block_height(ld->topology),
+			  feerate,
+			  type_to_string(tmpctx, struct amount_sat, &fee),
+			  type_to_string(tmpctx, struct amount_sat, &out_sats));
+
+		/* If we can afford fee and it's not dust, we're done */
+		if (amount_sat_sub(&amt, out_sats, fee)
+		    && amount_sat_greater_eq(amt, channel->our_config.dust_limit))
+			break;
+
+		/* Hmm, can't afford with recommended fee.  Try increasing deadline! */
+		block_target++;
+
+		/* If we can't even afford at FEERATE_FLOOR, something is wrong! */
+		if (feerate == FEERATE_FLOOR) {
+			amt = channel->our_config.dust_limit;
+			log_broken(channel->log, "TX can't afford minimal feerate"
+				   "; setting output to %s",
+				   type_to_string(tmpctx, struct amount_sat, &amt));
+			break;
+		}
+	}
+
+	/* We don't ever overflow here, right? */
+	assert(block_target + 20 > block_target);
+
+	/* If we anticipate waiting a long time, tell onchaind not to wait */
+	*worthwhile = (block_target < info->deadline_block + (u64)20);
```

Comment:
Can we give that `20` a name?

Reply:
`#define TWENTY 20`

Good point!

Reply:
Actually, I removed the bogus assertion, and now it only appears in one place.  I prefer to avoid indirection, so I simply added it to the comment:

```
 	/* If we anticipate waiting a long time (say, 20 blocks past
 	 * the deadline), tell onchaind not to wait */
```